folding-schemes/src/folding/nova/decider.rs

```diff
@@ -200,76 +196,49 @@ where
+            .map(|((v, _), &c)| {
+                CS1::prove_with_challenge(&pp.c1_cs_pp, c, v, &C1::ScalarField::zero(), None)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let c2_kzg_challenges = circuit2.kzg_challenges.clone();
+        let c2_kzg_proofs = circuit2
+            .cf_W_i
+            .get_openings()
+            .iter()
+            .zip(&c2_kzg_challenges)
+            .map(|((v, _), &c)| {
+                CS2::prove_with_challenge(&pp.c2_cs_pp, c, v, &C2::ScalarField::zero(), None)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let c1_snark_proof = S1::prove(&pp.c1_snark_pp, circuit1, &mut rng)
             .map_err(|e| Error::Other(e.to_string()))?;
-        let c2_snark_proof = S2::prove(&pp.c2_snark_pp, circuit2.clone(), &mut rng)
+        let c2_snark_proof = S2::prove(&pp.c2_snark_pp, circuit2, &mut rng)
             .map_err(|e| Error::Other(e.to_string()))?;
 
-        let cmT = circuit1.cmT.unwrap();
-        let W_i1 = circuit1.W_i1.unwrap();
-        let cf_W_i = circuit2.cf_W_i.unwrap();
-
-        // get the challenges that have been already computed when preparing the circuits inputs in
-        // the above `from_nova` calls
-        let challenge_W = circuit1
-            .cs_c_W
-            .ok_or(Error::MissingValue("cs_c_W".to_string()))?;
-        let challenge_E = circuit1
-            .cs_c_E
-            .ok_or(Error::MissingValue("cs_c_E".to_string()))?;
-        let c2_challenge_W = circuit2
-            .cs_c_W
-            .ok_or(Error::MissingValue("c2's cs_c_W".to_string()))?;
-        let c2_challenge_E = circuit2
-            .cs_c_E
-            .ok_or(Error::MissingValue("c2's cs_c_E".to_string()))?;
-
-        // generate CommitmentScheme proofs for the main instance
-        let U_cmW_proof = CS1::prove_with_challenge(
-            &pp.c1_cs_pp,
-            challenge_W,
-            &W_i1.W,
-            &C1::ScalarField::zero(),
-            None,
-        )?;
-        let U_cmE_proof = CS1::prove_with_challenge(
-            &pp.c1_cs_pp,
-            challenge_E,
-            &W_i1.E,
-            &C1::ScalarField::zero(),
-            None,
-        )?;
-        // CS proofs for the CycleFold instance
-        let cf_cmW_proof = CS2::prove_with_challenge(
-            &pp.c2_cs_pp,
-            c2_challenge_W,
-            &cf_W_i.W,
-            &C2::ScalarField::zero(),
-            None,
-        )?;
-        let cf_cmE_proof = CS2::prove_with_challenge(
-            &pp.c2_cs_pp,
-            c2_challenge_E,
-            &cf_W_i.E,
-            &C2::ScalarField::zero(),
-            None,
-        )?;
-
         Ok(Self::Proof {
             c1_snark_proof,
             c2_snark_proof,
-            cs1_proofs: [U_cmW_proof, U_cmE_proof],
-            cs2_proofs: [cf_cmW_proof, cf_cmE_proof],
+            cs1_proofs: c1_kzg_proofs.try_into().unwrap(),
+            cs2_proofs: c2_kzg_proofs.try_into().unwrap(),
```

Comment:
Great suggestion!

With https://github.com/privacy-scaling-explorations/sonobe/pull/145/commits/5c6e3c50ec61d1ce7d7771b6676506fe186e5dbc, most of `unwrap`s are replaced by safe alternatives, and explicit reasons are added to those that will never fail. Some methods in `solidity-verifiers` and unit tests still contain `unwrap` , which I think is fine :)

Reply:
Oh wow! you went much further and fixed them in all the other places & modules, awesome! Thanks! ^^
Yeah, unit tests for now are fine with unwraps. Is mainly for the lib methods, so that the lib does not panic on an error without the dev using it having the control.